src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs

```diff
@@ -129,22 +136,80 @@ private void OnRpcDisconnected(object sender, JsonRpcDisconnectedEventArgs e)
 
                 try
                 {
-                    var descriptor = new ServiceDescriptor(serviceName) { HostGroup = hostGroup };
                     return await client.RequestServiceAsync(descriptor, cancellationToken).ConfigureAwait(false);
                 }
                 catch (RemoteInvocationException ex)
                 {
-                    // RequestServiceAsync should never fail unless service itself is actually broken.
-                    // right now, we know only 1 case where it can randomly fail. but there might be more cases so 
-                    // adding non fatal watson here.
-                    FatalError.ReportWithoutCrash(ex);
+                    // save info only if it failed with different issue than before.
+                    if (lastException?.Message != ex.Message)
+                    {
+                        // RequestServiceAsync should never fail unless service itself is actually broken.
+                        // So far, we catched multiple issues from this NFW. so we will keep this NFW.
+                        WatsonReporter.Report("RequestServiceAsync Failed", ex, ReportDetailInfo);
+
+                        lastException = ex;
+                    }
                 }
 
                 // wait for retry_delayInMS before next try
                 await Task.Delay(retry_delayInMS, cancellationToken).ConfigureAwait(false);
             }
 
-            return Contract.FailWithReturn<Stream>("Fail to get service. look FatalError.s_reportedException for more detail");
+            // crash right away to get better dump. otherwise, we will get dump from async exception
+            // which most likely lost all valuable data
+            FatalError.ReportUnlessCanceled(lastException);
+            GC.KeepAlive(lastException);
+
+            // unreachable
+            throw ExceptionUtilities.Unreachable;
```

Comment:
Is this really unreachable? The call above it is `ReportUnlessCanceled` -- the name doesn't suggest you can't get past that point.